File: app/sprache.py
# -*- coding: utf-8 -*-
"""Welche Sprache PaperTree spricht.

PaperTree hat keine eigene Spracheinstellung: es nimmt die, die der Benutzer
in Paperless gewählt hat. Zwei Einstellungen für dieselbe Sache laufen sonst
auseinander, und niemand sucht sie an zwei Orten.

Hier stehen nur die beiden Umrechnungen von einem Wert aus `ui_settings` auf
das, was die Oberfläche braucht. Sie hängen an keiner Bibliothek – deshalb
lassen sie sich ohne laufendes Paperless prüfen.
"""
from __future__ import annotations

import json
import os
import re

# Die Oberfläche gibt es in diesen Sprachen; alles andere bekommt Englisch.
SPRACHEN = ("de", "en", "fr", "it", "es")

# Wo Paperless die beiden Werte ablegt. Die Sprache steht zuoberst, weil
# Paperless sie serverseitig selbst braucht.
SCHLUESSEL_SPRACHE = "language"

# Die Datumssprache liegt verschachtelt: settings.date_display.date_locale.
# Im Browser heisst dieselbe Einstellung "general-settings:date-display:
# date-locale" – das ist der Name im Frontend, beim Speichern baut Paperless
# daraus die verschachtelte Form. Nachgeschaut wird an beiden Orten, damit es
# auch mit einer Fassung stimmt, die flach ablegt.
SCHLUESSEL_DATUM = ("date_display", "date_locale")
SCHLUESSEL_DATUM_FLACH = "general-settings:date-display:date-locale"

# Neben den Sprachen kennt Paperless einen Sonderwert: ISO 8601. Das ist
# keine Sprache, sondern eine Schreibweise – 2026-09-24, überall gleich.
# Intl nimmt "iso-8601" als wohlgeformtes Sprachkürzel an und übergeht es
# dann stillschweigend, darum muss es hier von Hand durchgereicht werden.
ISO = "iso-8601"

# Ein Sprachkürzel, wie Intl es versteht: "de", "de-CH", "pt-BR".
_LOCALE = re.compile(r"^[A-Za-z]{2,3}(?:-[A-Za-z0-9]{2,8})*$")


def sprache_waehlen(wert: str | None) -> str:
    """Aus "de-de" wird "de"; was PaperTree nicht kann, wird zu "".

    Leer heisst nicht Englisch, sondern "keine Angabe" – dann entscheidet der
    Browser, genau wie Paperless es ohne gesetzte Sprache tut.
    """
    kurz = (wert or "").strip().lower().replace("_", "-").split("-")[0]
    return kurz if kurz in SPRACHEN else ""


def datumssprache_waehlen(wert: str | None) -> str:
    """Das Sprachkürzel fürs Datum, unverkürzt – "de-CH" schreibt anders als "de-DE".

    Anders als bei der Oberfläche gibt es hier keine Liste: das Datum
    formatiert der Browser, und der kennt jede Sprache. Geprüft wird nur die
    Form, damit aus den Einstellungen nichts Unbrauchbares weitergereicht wird.
    "iso-8601" kommt durch, obwohl es keine Sprache ist.
    """
    kurz = (wert or "").strip().replace("_", "-")
    if kurz.lower() == ISO:
        return ISO
    return kurz if _LOCALE.match(kurz) else ""


def _datumswert(werte: dict) -> str:
    """Die Datumseinstellung, verschachtelt oder flach."""
    aussen, innen = SCHLUESSEL_DATUM
    verschachtelt = werte.get(aussen)
    if isinstance(verschachtelt, dict) and verschachtelt.get(innen):
        return str(verschachtelt[innen])
    return str(werte.get(SCHLUESSEL_DATUM_FLACH) or "")


def aus_einstellungen(gewaehlt: dict | None) -> dict:
    """Sprache und Datumssprache aus dem settings-Teil von ui_settings.

    Fürs Datum nennt Paperless eine eigene Sprache und fällt sonst auf die
    Anzeigesprache zurück. PaperTree macht es genauso, damit dasselbe
    Dokument in beiden Oberflächen dasselbe Datum zeigt.
    """
    werte = gewaehlt or {}
    roh_sprache = werte.get(SCHLUESSEL_SPRACHE) or ""
    return {
        "sprache": sprache_waehlen(roh_sprache),
        "datumssprache": (datumssprache_waehlen(_datumswert(werte))
                          or datumssprache_waehlen(roh_sprache)),
    }


# --- Texte für die wenigen Seiten, die der Server selbst schreibt -----------
# Sonst kommt jeder Text aus dem Browser. Zwei Fälle gehen aber direkt an den
# Benutzer, ohne dass die Oberfläche dazwischenliegt: die Vorschau im Rahmen
# und der Download. Beide holen ihre Texte hier – aus denselben Dateien wie
# der Rest, damit es nicht zwei Quellen für dieselbe Formulierung gibt.
_SPRACHORDNER = os.path.join(
    os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "web", "sprachen")
_KATALOGE: dict = {}


def _katalog(sprache: str) -> dict:
    if sprache not in _KATALOGE:
        pfad = os.path.join(_SPRACHORDNER, sprache + ".json")
        try:
            with open(pfad, encoding="utf-8") as datei:
                _KATALOGE[sprache] = json.load(datei)
        except (OSError, ValueError):
            _KATALOGE[sprache] = {}
    return _KATALOGE[sprache]


def text(sprache: str, schluessel: str) -> str:
    """Ein Text in der Sprache des Benutzers, sonst auf Englisch.

    Fehlt er in beiden, kommt der Schlüssel zurück – das fällt auf, statt
    stillschweigend eine leere Stelle zu hinterlassen.
    """
    gewaehlt = _katalog(sprache_waehlen(sprache) or "en")
    if schluessel in gewaehlt:
        return gewaehlt[schluessel]
    return _katalog("en").get(schluessel, schluessel)

File: web/js/sprache.js
// Die Sprache der Oberfläche.
//
// PaperTree hat keine eigene Spracheinstellung. Es spricht die Sprache, die
// der Benutzer in Paperless gewählt hat – zwei Einstellungen für dieselbe
// Sache laufen sonst auseinander, und niemand sucht sie an zwei Orten.
//
// Fünf Sprachen liegen bei. Wer Paperless auf etwas anderes gestellt hat,
// bekommt Englisch; Englisch ist auch der Rückfall für einen Schlüssel, den
// eine Übersetzung noch nicht kennt. Deshalb liegt der englische Katalog
// immer mit im Speicher: eine fehlende Zeile soll einen einzelnen Text
// betreffen und nicht die halbe Ansicht.
import { BASIS } from './api.js';

export const SPRACHEN = ['de', 'en', 'fr', 'it', 'es'];
export const RUECKFALL = 'en';

// Paperless kennt fürs Datum neben den Sprachen einen Sonderwert: ISO 8601,
// also 2026-09-24, in jeder Sprache gleich. Intl nimmt "iso-8601" als
// wohlgeformtes Sprachkürzel an und übergeht es dann stillschweigend – wer
// das einstellt, sähe sonst weiter das Datum seiner Anzeigesprache.
export const ISO = 'iso-8601';

let katalog = {};
let englisch = {};
let aktuelleSprache = RUECKFALL;
let datumsSprache = RUECKFALL;

// Intl-Formatierer sind teuer genug, dass es sich lohnt, sie zu behalten.
let datumsformat = null;
let zahlformat = null;
let pluralregel = null;

/** Der Sprachcode, in dem gerade gesprochen wird. */
export const sprache = () => aktuelleSprache;

/** Das Sprachkürzel, mit dem Datum und Zahlen geschrieben werden. */
export const datumssprache = () => datumsSprache;

function katalogHolen(code) {
  return fetch(BASIS + 'sprachen/' + code + '.json', { credentials: 'same-origin' })
    .then((antwort) => (antwort.ok ? antwort.json() : {}))
    .catch(() => ({}));
}

/** Welche Sprache gilt: die aus Paperless, sonst die des Browsers, sonst Englisch. */
export function spracheBestimmen(ausPaperless) {
  if (SPRACHEN.includes(ausPaperless)) return ausPaperless;
  for (const wunsch of navigator.languages || [navigator.language || '']) {
    const kurz = String(wunsch).toLowerCase().split('-')[0];
    if (SPRACHEN.includes(kurz)) return kurz;
  }
  return RUECKFALL;
}

/** Prüft, ob der Browser dieses Sprachkürzel fürs Datum kennt. */
function datumsspracheBestimmen(gewuenscht, ersatz) {
  if (gewuenscht === ISO) return ISO;
  for (const kandidat of [gewuenscht, ersatz]) {
    if (!kandidat || kandidat === ISO) continue;
    try {
      // Wirft bei unbrauchbarer Angabe – dann gilt der nächste Kandidat.
      new Intl.DateTimeFormat(kandidat);
      return kandidat;
    } catch (_) { /* nächster */ }
  }
  return RUECKFALL;
}

/**
 * Lädt die Kataloge und stellt die Formatierer ein.
 *
 * `benutzer` ist die Auskunft aus /api/ich: sie nennt die Sprache aus
 * Paperless und, getrennt davon, die Sprache fürs Datum.
 */
export async function spracheEinrichten(benutzer) {
  const person = benutzer || {};
  aktuelleSprache = spracheBestimmen(person.sprache || '');
  datumsSprache = datumsspracheBestimmen(person.datumssprache || '', aktuelleSprache);

  const [gewaehlt, ersatz] = await Promise.all([
    katalogHolen(aktuelleSprache),
    aktuelleSprache === RUECKFALL ? Promise.resolve(null) : katalogHolen(RUECKFALL),
  ]);
  katalog = gewaehlt || {};
  englisch = ersatz || katalog;

  // ISO ist keine Sprache: Zahlen schreibt dann die Anzeigesprache, sonst
  // stünde in einer deutschen Oberfläche plötzlich 1,660 statt 1'660.
  const zahlSprache = datumsSprache === ISO ? aktuelleSprache : datumsSprache;
  datumsformat = datumsSprache === ISO
    ? null : new Intl.DateTimeFormat(datumsSprache, { dateStyle: 'medium' });
  zahlformat = new Intl.NumberFormat(zahlSprache);
  pluralregel = new Intl.PluralRules(aktuelleSprache);

  // Damit der Browser richtig trennt und vorliest.
  document.documentElement.lang = aktuelleSprache;
  return aktuelleSprache;
}

function einsetzen(vorlage, werte) {
  if (!werte) return vorlage;
  return vorlage.replace(/\{(\w+)\}/g, (ganz, name) =>
    (Object.prototype.hasOwnProperty.call(werte, name) ? String(werte[name]) : ganz));
}

/**
 * Der Text zu einem Schlüssel.
 *
 * Fehlt er in beiden Katalogen, kommt der Schlüssel selbst zurück. Das ist
 * hässlich und genau deshalb richtig: die Lücke fällt auf, statt dass die
 * Stelle leer bleibt und niemand sie bemerkt.
 */
export function t(schluessel, werte) {
  let text = katalog[schluessel];
  if (text === undefined) text = englisch[schluessel];
  if (text === undefined) return schluessel;
  return einsetzen(text, werte);
}

/**
 * Ein Text mit Anzahl: `tn('suche.treffer', 3)` sucht sich zwischen
 * `suche.treffer.one` und `suche.treffer.other` das Richtige.
 *
 * Welche Form gilt, sagt der Browser. Im Französischen gehört die Null zur
 * Einzahl, im Deutschen nicht – das steht in keiner Tabelle hier, sondern in
 * Intl.PluralRules.
 */
export function tn(schluessel, anzahl, werte) {
  const zahl = Number(anzahl) || 0;
  const form = pluralregel ? pluralregel.select(zahl) : (zahl === 1 ? 'one' : 'other');
  const voll = { anzahl: zahlText(zahl), ...(werte || {}) };
  let text = katalog[schluessel + '.' + form];
  if (text === undefined) text = katalog[schluessel + '.other'];
  if (text === undefined) text = englisch[schluessel + '.' + form];
  if (text === undefined) text = englisch[schluessel + '.other'];
  if (text === undefined) return schluessel;
  return einsetzen(text, voll);
}

/** Ob es diesen Schlüssel gibt – für Fälle, die ohne Text auskommen. */
export const kenntText = (schluessel) =>
  katalog[schluessel] !== undefined || englisch[schluessel] !== undefined;

function isoText(zeitpunkt) {
  const zwei = (n) => String(n).padStart(2, '0');
  return zeitpunkt.getFullYear() + '-' + zwei(zeitpunkt.getMonth() + 1)
    + '-' + zwei(zeitpunkt.getDate());
}

/**
 * Ein Datum in der mittleren Form: "24. Sept. 2026", "Sep 24, 2026" – oder
 * "2026-09-24", wenn in Paperless ISO 8601 eingestellt ist.
 *
 * Immer mittel, nie ausgeschrieben – in einer Liste mit dreissig Zeilen zählt
 * die gleiche Breite mehr als der volle Monatsname, und die kurze Form
 * schreibt das Jahr zweistellig, was in einer Ablage niemand will.
 *
 * Ein reiner Datumstext wird von Hand zerlegt statt durch Date geschickt:
 * "2026-01-23" wäre dort UTC-Mitternacht und fiele westlich von Greenwich auf
 * den Vortag.
 */
export function datum(wert) {
  if (!wert) return '';
  const text = String(wert);
  const iso = /^(\d{4})-(\d{2})-(\d{2})/.exec(text);
  const zeitpunkt = iso
    ? new Date(Number(iso[1]), Number(iso[2]) - 1, Number(iso[3]))
    : new Date(text);
  if (isNaN(zeitpunkt)) return text.slice(0, 10);
  if (datumsSprache === ISO) return isoText(zeitpunkt);
  return datumsformat ? datumsformat.format(zeitpunkt) : isoText(zeitpunkt);
}

/** Eine Zahl mit den Trennzeichen der eingestellten Sprache. */
export function zahlText(wert) {
  const zahl = Number(wert);
  if (!isFinite(zahl)) return String(wert == null ? '' : wert);
  return zahlformat ? zahlformat.format(zahl) : String(zahl);
}

/**
 * Setzt die festen Texte im HTML.
 *
 * `data-i18n` ersetzt den Inhalt, `data-i18n-titel`, `data-i18n-marke` und
 * `data-i18n-platzhalter` die zugehörigen Attribute. So bleibt index.html
 * lesbar: dort steht weiter, was gemeint ist, und daneben der Schlüssel.
 */
export function textenSetzen(wurzel = document) {
  for (const knoten of wurzel.querySelectorAll('[data-i18n]')) {
    knoten.textContent = t(knoten.getAttribute('data-i18n'));
  }
  const attribute = [
    ['data-i18n-titel', 'title'],
    ['data-i18n-marke', 'aria-label'],
    ['data-i18n-platzhalter', 'placeholder'],
  ];
  for (const [quelle, ziel] of attribute) {
    for (const knoten of wurzel.querySelectorAll('[' + quelle + ']')) {
      knoten.setAttribute(ziel, t(knoten.getAttribute(quelle)));
    }
  }
}
